libbindgen/src/ir/context.rs

```diff
@@ -441,6 +443,40 @@ impl<'ctx> BindgenContext<'ctx> {
         ret
     }
 
+    /// This function trying to find any dangling references inside of `items`
+    fn assert_no_dangling_references(&self) {
+        if cfg!(debug_assertions) {
+            for (id, item) in self.items() {
+                let mut sub_items = ItemSet::new();
+                item.collect_types(self, &mut sub_items, &());
+
+                sub_items.insert(item.parent_id());
+                for sub in sub_items {
+                    assert!(self.resolve_item_fallible(sub).is_some(),
+                            "Should not dangle: {:?} references {:?}", id, sub);
+                }
+            }
+        }
+    }
+
+    fn related_items_traversal<'me>(&'me self)
+                                    -> RelatedItemsIter<'me, 'ctx> {
+        assert!(self.in_codegen_phase());
+        assert!(self.current_module == self.root_module);
+
+        let roots = self.items().map(|(&id, _)| id);
+
+        let seen: ItemSet = roots.collect();
```

Comment:
I think you want to keep the `seen` set empty, and also put the `self.items()` iterator as a member into `RelatedItemsIter`.